pkg/scheduler/schedule_one_test.go

```diff
@@ -448,6 +487,136 @@ func TestSchedulerMultipleProfilesScheduling(t *testing.T) {
+					QueueSort: schedulerapi.PluginSet{Enabled: []schedulerapi.Plugin{{Name: "PrioritySort"}}},
+					Bind:      schedulerapi.PluginSet{Enabled: []schedulerapi.Plugin{{Name: "DefaultBinder"}}},
+				},
+			},
+		),
+		WithFrameworkOutOfTreeRegistry(frameworkruntime.Registry{
+			"FakeNodeSelectorDependOnPodAnnotation": newFakeNodeSelectorDependOnPodAnnotation,
+		}),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Run scheduler.
+	informerFactory.Start(ctx.Done())
+	informerFactory.WaitForCacheSync(ctx.Done())
+	go sched.Run(ctx)
+
+	// following we start 2 goroutines asynchronously to detect potential racing issues
+	// 1) one is responsible for deleting several nodes in each round
+	// 2) another is creating several pods in each round to trigger scheduling
+	// those two goroutines will stop until ctx.Done() is called, which means all waiting pods are scheduled at least once
+	var deleteNodeIndex int
+	deleteNodesOneRound := func() {
+		for i := 0; i < deleteNodeNumberPreRound; i++ {
+			if deleteNodeIndex >= initialNodeNumber {
+				// all initial nodes are already deleted
+				return
+			}
+			deleteNodeName := fmt.Sprintf("node%d", deleteNodeIndex)
+			if err := client.CoreV1().Nodes().Delete(ctx, deleteNodeName, metav1.DeleteOptions{}); err != nil {
+				t.Fatal(err)
+			}
+			deleteNodeIndex++
+		}
+	}
+	go wait.Until(deleteNodesOneRound, 15*time.Millisecond, ctx.Done())
+
+	// creating pods in each round
+	var createPodIndex int
+	createPodsOneRound := func() {
+		if createPodIndex > waitSchedulingPodNumber {
+			return
+		}
+		for i := 0; i < createPodNumberPreRound; i++ {
+			podName := fmt.Sprintf("pod%d", createPodIndex)
+			// note: the node(specifiedNodeName) may already be deleted, which leads pod scheduled failed
+			specifiedNodeName := fmt.Sprintf("node%d", random.Intn(initialNodeNumber))
+
+			waitSchedulingPod := st.MakePod().Namespace(fakeNamespace).Name(podName).UID(podName).Annotation(fakeSpecifiedNodeNameAnnotation, specifiedNodeName).SchedulerName(fakeSchedulerName).Obj()
+			if _, err := client.CoreV1().Pods(fakeNamespace).Create(ctx, waitSchedulingPod, metav1.CreateOptions{}); err != nil {
+				t.Fatal(err)
+			}
+			createPodIndex++
+		}
+	}
+	go wait.Until(createPodsOneRound, 14*time.Millisecond, ctx.Done())
+
+	// capture the events to wait all pods to be scheduled at least once
+	allWaitSchedulingPods := sets.NewString()
+	for i := 0; i < waitSchedulingPodNumber; i++ {
+		allWaitSchedulingPods.Insert(fmt.Sprintf("pod%d", i))
+	}
+	var wg sync.WaitGroup
+	wg.Add(waitSchedulingPodNumber)
+	stopFn, err := broadcaster.StartEventWatcher(func(obj runtime.Object) {
+		e, ok := obj.(*eventsv1.Event)
+		if !ok || (e.Reason != "Scheduled" && e.Reason != "FailedScheduling") {
+			return
+		}
+		if allWaitSchedulingPods.Has(e.Regarding.Name) {
+			wg.Done()
+			allWaitSchedulingPods.Delete(e.Regarding.Name)
+		}
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer stopFn()
+
+	wg.Wait()
```

Comment:
This is logically accurate. But it seems it'd make this UT a bit slow:

```
$ go test ./pkg/scheduler -run TestSchedulerGuaranteeNonNilNodeInSchedulingCycle
ok  	k8s.io/kubernetes/pkg/scheduler	1.771s
```

I don't think a single UT would cost ~2s is a good idea... Do you think if we can achieve the same goal by simply plumbing a counter (make it atomic and plus 1 for each call) in the fake plugin, and check the completion on the counter?

Reply:
Yes, the execution time of this UT should be optimized. And I found that most of the time was spent waiting for the func `createPodsOneRound` to be called, as calculated by the following:

```
(waitSchedulingPodNumber / createPodNumberPerRound) * waitPeriod = (2000 / 25) * 14ms = 1.12s
```

Then, I update those parameters as following:
- `waitSchedulingPodNumber` from 2000 to 200
- `createPodNumberPerRound` from 25 to 50
- waitPeriod from 14ms to 10ms

```
(waitSchedulingPodNumber / createPodNumberPerRound) * waitPeriod = (200 / 50) * 10ms = 0.04s
```

And, the overall UT time consumption on my Mac is `0.19s`
```
=== RUN   TestSchedulerGuaranteeNonNilNodeInSchedulingCycle
--- PASS: TestSchedulerGuaranteeNonNilNodeInSchedulingCycle (0.19s)
PASS
```

In this case, I am thinking of keeping the fake plugin as it is now~